From ae49ae3f657dc0086eca8873213cf9dcee953ab8 Mon Sep 17 00:00:00 2001
From: John Ralls <jralls@ceridwen.us>
Date: Thu, 12 Jun 2025 16:06:02 -0700
Subject: [PATCH] [Quartz] Set the view's frame in a move for temp windows.

Temp windows (context menus mostly) are created at 0,0 then moved to the
pointer location. When the move is from a Retina to a non-Retina monitor
the view needs to be redrawn at the new scale, but macOS doesn't recognize
that and the menu ends up truncated. Setting the frame on the view forces
the redraw.

Fixes https://gitlab.gnome.org/GNOME/gtk/-/issues/7598
---
 gdk/quartz/gdkwindow-quartz.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/gdk/quartz/gdkwindow-quartz.c b/gdk/quartz/gdkwindow-quartz.c
index 9a6aea9c6f..08e47dd089 100644
--- a/gdk/quartz/gdkwindow-quartz.c
+++ b/gdk/quartz/gdkwindow-quartz.c
@@ -1393,6 +1393,10 @@ move_resize_window_internal (GdkWindow *window,
 
       frame_rect = [impl->toplevel frameRectForContentRect:content_rect];
       [impl->toplevel setFrame:frame_rect display:YES];
+
+      if (window->window_type == GDK_WINDOW_TEMP)
+        [impl->view setFrame:content_rect];
+
       impl->cairo_surface = gdk_quartz_ref_cairo_surface (window);
       cairo_surface_destroy (impl->cairo_surface); // Remove the extra reference
     }
-- 
2.39.5 (Apple Git-154)

